// fortnite-client/lib/models/store/catalog-entry.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const class_transformer_1 = require("class-transformer");
const gift_info_1 = require("./gift-info");
const item_grant_1 = require("./item-grant");
const meta_asset_info_1 = require("./meta-asset-info");
const meta_info_1 = require("./meta-info");
const price_1 = require("./price");
const requirement_1 = require("./requirement");
class CatalogEntry {
    static FROM_JSON(jsonObject) {
        return class_transformer_1.plainToClass(CatalogEntry, jsonObject);
    }
    /* istanbul ignore next */
    toJson() {
        return class_transformer_1.classToPlain(this);
    }
}
__decorate([
    class_transformer_1.Type(() => price_1.Price),
    __metadata("design:type", Array)
], CatalogEntry.prototype, "prices", void 0);
__decorate([
    class_transformer_1.Type(() => requirement_1.Requirement),
    __metadata("design:type", Array)
], CatalogEntry.prototype, "requirements", void 0);
__decorate([
    class_transformer_1.Type(() => meta_info_1.MetaInfo),
    __metadata("design:type", Array)
], CatalogEntry.prototype, "metaInfo", void 0);
__decorate([
    class_transformer_1.Type(() => item_grant_1.ItemGrant),
    __metadata("design:type", Array)
], CatalogEntry.prototype, "itemGrants", void 0);
__decorate([
    class_transformer_1.Type(() => gift_info_1.GiftInfo),
    __metadata("design:type", gift_info_1.GiftInfo)
], CatalogEntry.prototype, "giftInfo", void 0);
__decorate([
    class_transformer_1.Type(() => meta_asset_info_1.MetaAssetInfo),
    __metadata("design:type", meta_asset_info_1.MetaAssetInfo)
], CatalogEntry.prototype, "metaAssetInfo", void 0);
exports.CatalogEntry = CatalogEntry;
var CatalogGroup;
(function (CatalogGroup) {
    CatalogGroup["Empty"] = "";
})(CatalogGroup = exports.CatalogGroup || (exports.CatalogGroup = {}));
var OfferType;
(function (OfferType) {
    OfferType["StaticPrice"] = "StaticPrice";
})(OfferType = exports.OfferType || (exports.OfferType = {}));
